src/Stratis.Features.SQLiteWalletRepository/SQLiteWalletRepositoryExt.cs

```diff
@@ -0,0 +1,128 @@
+
+        internal static TransactionData ToTransactionData(this SQLiteWalletRepository repo, HDTransactionData transactionData, IEnumerable<HDPayment> payments)
+        {
+            return new TransactionData()
+            {
+                Amount = new Money(transactionData.Value, MoneyUnit.BTC),
+                BlockHash = uint256.Parse(transactionData.OutputBlockHash),
+                BlockHeight = transactionData.OutputBlockHeight,
+                CreationTime = DateTimeOffset.FromUnixTimeSeconds(transactionData.OutputTxTime),
+                Id = uint256.Parse(transactionData.OutputTxId),
+                Index = transactionData.OutputIndex,
+                // These two are always updated and used in tandem so we update them from a single source value.
+                IsCoinBase = transactionData.OutputTxIsCoinBase == 1,
+                IsCoinStake = transactionData.OutputTxIsCoinBase == 1,
+                // IsPropagated  // Not used currently.
+                ScriptPubKey = new Script(Encoders.Hex.DecodeData(transactionData.RedeemScript)),
+                SpendingDetails = (transactionData.SpendTxId == null) ? null : new SpendingDetails()
+                {
+                    BlockHeight = transactionData.SpendBlockHeight,
+                    // BlockIndex // Not used currently.
+                    CreationTime = DateTimeOffset.FromUnixTimeSeconds((int)transactionData.SpendTxTime),
+                    IsCoinStake = transactionData.SpendTxIsCoinBase == 1,
+                    TransactionId = uint256.Parse(transactionData.SpendTxId),
+                    Payments = payments.Select(p => new PaymentDetails()
+                    {
+                         Amount = new Money((decimal)p.SpendValue, MoneyUnit.BTC),
+                         DestinationScriptPubKey = new Script(Encoders.Hex.DecodeData(p.SpendScriptPubKey)),
+                         OutputIndex = p.SpendIndex
+                    }).ToList()
+                }
+            };
+        }
+
+        internal static IEnumerable<Script> GetDestinations(this SQLiteWalletRepository repo, Script redeemScript)
+        {
+            ScriptTemplate scriptTemplate = repo.Network.StandardScriptsRegistry.GetTemplateFromScriptPubKey(redeemScript);
+
+            if (scriptTemplate != null)
+            {
+                // We need scripts suitable for matching to HDAddress.ScriptPubKey.
+                switch (scriptTemplate.Type)
+                {
+                    case TxOutType.TX_PUBKEYHASH:
+                        yield return redeemScript;
+                        break;
+                    case TxOutType.TX_PUBKEY:
+                        yield return PayToPubkeyTemplate.Instance.ExtractScriptPubKeyParameters(redeemScript).Hash.ScriptPubKey;
+                        break;
+                    case TxOutType.TX_SCRIPTHASH:
+                        yield return PayToScriptHashTemplate.Instance.ExtractScriptPubKeyParameters(redeemScript).ScriptPubKey;
```

Comment:
Is the design intended to accommodate P2SH transactions as well? Their scripts don't map well into an HD structure as they are essentially arbitrary.

Reply:
I am making the minimum assumption that a script can always be mapped to one or more addresses, otherwise how would we add it to a wallet?